lib/managedservicesplatform/runtime/job.go

```diff
@@ -0,0 +1,130 @@
+package runtime
+
+import (
+	"context"
+	"flag"
+	"os"
+
+	"cloud.google.com/go/profiler"
+	"github.com/sourcegraph/log"
+
+	"github.com/sourcegraph/sourcegraph/lib/managedservicesplatform/runtime/contract"
+	"github.com/sourcegraph/sourcegraph/lib/managedservicesplatform/runtime/internal/opentelemetry"
+)
+
+type Job[ConfigT any] interface {
+	contract.ServiceMetadataProvider
+	// Execute should use given configuration to perform the desired
+	// job execution.
+	Execute(
+		ctx context.Context,
+		logger log.Logger,
+		contract Contract,
+		config ConfigT,
+	) error
+}
+
+// ExecuteJob handles the entire lifecycle of the program performing a
+// job execution, and should be the only thing called in a MSP job
+// program's main package, for example:
+//
+//	runtime.ExecuteJob[example.Config](example.Job{})
+//
+// Where example.Config is your runtime.ConfigLoader implementation, and
+// example.Job is your runtime.Job implementation.
+func ExecuteJob[
+	ConfigT any,
+	LoaderT ConfigLoader[ConfigT],
+](job Job[ConfigT]) {
+	// To get around os.Exit not honoring defer we
+	// wrap the body in another function which we can safely defer inside
+	err := func() (err error) {
```

Comment:
The control flow here is a little bit confusing; all error paths except the actual execution use `logger.Fatal` to exit. Maybe we can do:

```suggestion
	// To get around os.Exit not honoring defer we
	// wrap the body in another function which we can safely defer inside
	failed := func() (failed bool) {
```

Instead?

Reply:
oh 🤦 
ofc logger.Fatal calls os.Exit

Reply:
I think the structure here still makes sense, as the error logging is handled by job execution check-in - we just want to let defers handle properly before using os.Exit ourselves. logger.Fatal isn't as bad because it will at least flush logs correctly first

Reply:
(but, maybe we should refactor both entrypoints to avoid logger.Fatal so that full cleanup can happen 🤔 )

Reply:
Given that `defer done(&err)` wasn't adding any benefit I remove the `defer` and moved the call to `done` after `job.Execute`. As such we don't need a nested function anymore.

The `log.Fatal` calls are still a concern as we might never reach the point to check in. I think that might be okay though as like you said it will flush the logs first